# 제로베이스 데이터 사이언스 스쿨 3기 SQL 과제 2
- 유가 데이터 분석

### 8문제 총 100점

- 1번 5점
- 2번 5점
- 3번 10점
- 4번 30점
- 5번 10점
- 6번 10점
- 7번 15점
- 8번 15점

In [132]:
import mysql.connector
import csv
import pandas as pd
import googlemaps as gmaps 

In [117]:
conn = mysql.connector.connect(
    host="localhost",
    user="root",
    password="suhjh081",
    database="zerobase"
)

### 1 ~ 8번 모두 본 노트북 파일에 답안 작성해서 제출해주세요 :)

---

문제 1.

주유소 데이터를 저장할 테이블을 다음의 구조로 생성하세요. (PDF 파일 참고)

In [6]:

cursor = conn.cursor()
# SQL for creating the GAS_BRAND table
sql_gas_brand = """
CREATE TABLE GAS_BRAND (
    id INT PRIMARY KEY AUTO_INCREMENT,
    name VARCHAR(16)
);
"""

# SQL for creating the GAS_STATION table
sql_gas_station = """
CREATE TABLE GAS_STATION (
    id INT PRIMARY KEY AUTO_INCREMENT,
    brand INT,
    name VARCHAR(64),
    city CHAR(2),
    gu VARCHAR(10),
    address VARCHAR(128),
    gasoline INT,
    diesel INT,
    self BOOLEAN,
    car_wash BOOLEAN,
    charging_station BOOLEAN,
    car_maintenance BOOLEAN,
    convenience_store BOOLEAN,
    24_hours BOOLEAN,
    lat DECIMAL(16,14),
    lng DECIMAL(17,14),
    FOREIGN KEY (brand) REFERENCES GAS_BRAND(id)
);
"""

# Execute the SQL
cursor.execute(sql_gas_brand)
cursor.execute(sql_gas_station)

# Close the connection
cursor.close()
conn.close()

문제 2.

Python 코드로 GAS_BRAND 데이터를 다음과 같이 입력하고 확인하세요. (PDF 파일 참고)

In [27]:
def fetch_gas_brand_data():
    cursor = conn.cursor()
    cursor.execute("SELECT * FROM GAS_BRAND;")
    result = cursor.fetchall()
    cursor.close()
    return result

In [29]:
gas_brand_data = fetch_gas_brand_data()

print("Data from GAS_BRAND:")
for row in gas_brand_data:
    print(row)

conn.close()

Data from GAS_BRAND:
(1, 'SK에너지')
(2, 'HD현대오일뱅크')
(3, 'GS칼텍스')
(4, 'S-OIL')
(5, '알뜰주유소')
(6, '알뜰(ex)')
(7, '자가상표')


제출 1.
- Table 생성 결과 : Desc GAS_BRAND; Desc GAS_STATION;

In [31]:
def describe_table(table_name):
    cursor = conn.cursor()
    cursor.execute(f"DESC {table_name};")
    result = cursor.fetchall()
    cursor.close()
    return result

In [33]:
gas_brand_structure = describe_table('GAS_BRAND')

print("GAS_BRAND Structure:")


gas_brand_structure 

for row in gas_brand_structure:
    print(row)

GAS_BRAND Structure:
('id', b'int', 'NO', 'PRI', None, 'auto_increment')
('name', b'varchar(16)', 'YES', '', None, '')


In [34]:
gas_station_structure = describe_table('GAS_STATION')

print("GAS_STATION Structure:")
for row in gas_station_structure:
    print(row)

conn.close()

GAS_STATION Structure:
('id', b'int', 'NO', 'PRI', None, 'auto_increment')
('brand', b'int', 'YES', 'MUL', None, '')
('name', b'varchar(64)', 'YES', '', None, '')
('city', b'char(2)', 'YES', '', None, '')
('gu', b'varchar(10)', 'YES', '', None, '')
('address', b'varchar(128)', 'YES', '', None, '')
('gasoline', b'int', 'YES', '', None, '')
('diesel', b'int', 'YES', '', None, '')
('self', b'tinyint(1)', 'YES', '', None, '')
('car_wash', b'tinyint(1)', 'YES', '', None, '')
('charging_station', b'tinyint(1)', 'YES', '', None, '')
('car_maintenance', b'tinyint(1)', 'YES', '', None, '')
('convenience_store', b'tinyint(1)', 'YES', '', None, '')
('24_hours', b'tinyint(1)', 'YES', '', None, '')
('lat', b'decimal(16,14)', 'YES', '', None, '')
('lng', b'decimal(17,14)', 'YES', '', None, '')


제출 2.
- GAS_BRAND 조회 결과 : SELECT * FROM GAS_BRAND;

In [118]:
sql_query_q2 = """
SELECT * FROM GAS_STATION limit 10;
"""
cur = conn.cursor()
cur.execute(sql_query_q2)
sql_result_q2 = cur.fetchall()
cur.close()

for row in sql_result_q2:
    print(row)


(1, 1, '(주)보성 세곡주유소', '서울', '강남구', '서울 강남구 헌릉로 731 (세곡동)', 1562, 1439, 1, 1, 0, 0, 0, 0, Decimal('37.46715650000000'), Decimal('127.11870810000000'))
(2, 1, '자곡셀프주유소', '서울', '강남구', '서울 강남구 밤고개로 120 (자곡동)', 1609, 1474, 1, 1, 0, 0, 0, 0, Decimal('37.48340020000000'), Decimal('127.10404580000000'))
(3, 2, 'HD현대오일뱅크㈜직영 산성셀프주유소', '서울', '강남구', '서울 강남구 헌릉로 730', 1609, 1464, 1, 1, 0, 0, 0, 0, Decimal('37.46656750000000'), Decimal('127.11908030000000'))
(4, 3, '방죽주유소', '서울', '강남구', '서울 강남구 밤고개로 215 (율현동)', 1617, 1499, 1, 1, 0, 0, 0, 0, Decimal('37.47499260000000'), Decimal('127.10688660000000'))
(5, 1, '오일프러스 셀프', '서울', '강남구', '서울 강남구 남부순환로 2651 (도곡동)', 1619, 1524, 1, 1, 0, 1, 0, 1, Decimal('37.48583900000000'), Decimal('127.03967210000000'))
(6, 4, '극동유화㈜ 개나리주유소', '서울', '강남구', '서울 강남구 언주로 423 (역삼동)', 1629, 1534, 1, 1, 0, 1, 0, 1, Decimal('37.50007760000000'), Decimal('127.03854190000000'))
(7, 2, 'HD현대오일뱅크㈜직영 도곡셀프주유소', '서울', '강남구', '서울 강남구 남부순환로 2718 (도곡2동)', 1629, 1554, 1, 1, 0, 1, 0, 0, Deci

문제 3.

다음의 함수와 그 함수를 테스트하는 코드를 작성하세요. (PDF 파일 참고)

- 화폐단위 문자형을 입력받아 숫자형으로 반환하는 함수 (테스트 입력 : ‘1,000’)
- 주유소 브랜드를 입력하면 GAS_BRAND 데이터를 참고하여 ID 를 반환하는 함수 (테스트 입력 : ‘SK에너지’) - 주소를 입력받아 구 이름을 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)
- 주소를 입력받아 위도, 경도를 반환하는 함수 (테스트 입력 : ‘서울시 강남구 헌릉로 730’)

제출 3.
- 각 함수 및 테스트 코드와 실행결과

In [109]:
def currency_to_int(currency):
    return int(currency.replace(',', '')) 

#I have used parameterized queries instead of string formatting to prevent SQL injection.

def get_id_by_brand(brand):
    cursor = conn.cursor()
    cursor.execute(f"SELECT id FROM GAS_BRAND WHERE name = %s;", (brand,)) 
    result = cursor.fetchone()
    cursor.close()
    return result[0]

def get_gu_by_address(address):
    cursor = conn.cursor()
    cursor.execute(f"SELECT gu FROM GAS_STATION WHERE address = %s;", (address,))
    result = cursor.fetchone()
    cursor.close()
    return result[0]

def get_lat_lng_by_address(address):
    cursor = conn.cursor()
    cursor.execute(f"SELECT lat, lng FROM GAS_STATION WHERE address = %s;", (address,)) 
    result = cursor.fetchone()
    cursor.close()
    return result


In [110]:
print(get_id_by_brand('SK에너지'))
print(get_gu_by_address('서울시 송파구 위례중앙로 10'))
print(get_lat_lng_by_address('서울시 송파구 위례중앙로 10'))


1
송파구
(Decimal('37.47542740000000'), Decimal('127.12847070000000'))


문제 4.

Python 코드에서 주유소 페이지에서 데이터를 가져올때, GAS_STATION 테이블에 바로 입력하도록 수정하세요. (앞서 생 성한 함수활용)
- 주의. city 는 ‘서울’ 로 고정, 부가정보 데이터 타입
- 입력된 데이터의 총 갯수를 쿼리하여 결과를 확인합니다. - 입력된 데이터 상위 10개를 쿼리하여 결과를 확인합니다.


제출 4.
- 주유소 데이터 관련 코드 (ipynb)



In [ ]:
from selenium import webdriver
from selenium.webdriver.common.by import By
url = 'https://www.opinet.co.kr/searRgSelect.do'
driver = webdriver.Chrome()
driver.get(url)
driver.get(url)

import time
from selenium.webdriver.support.wait import WebDriverWait

ql= "Insert Into GAS_STATION \
	(id, brand,name, city,gu, address, gasoline, diesel, \
     self,car_wash,charging_station, car_maintenance, \
     convenience_store, 24_hours, lat, lng) \
    VALUES (%s, %s, %s, '서울', %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s);"

# CSS_SELECTOR를 입력하면 src 속성값을 반환하는 함수
def getSRC(selector):
	return driver.find_elemet(By.CSS_SELECTOR, selector).get_attrubute('src')
    
# ,를 포함하는 숫자가 문자형으로 입력되면 숫자로 변환하는 함수
def str_to_int(string):
    return int(string.replace(',',''))

# address가 입력되면 lat,lng 반환하는 함수
import googlemaps
gmaps_key = "AIzaSyD-4X5qQ8X0Z3Q3Z3Q3Z3Q3Z3Q3Z3Q3Z3Q"
gmaps = googlemaps.Client(key = gmaps_key)

def address_to_latlng(address):
    tmp = gmaps.geocode(address, language ="ko")
    lat = tmp[0].get("geometry")["location"]["lat"]
    lng = tmp[0].get("geometry")["location"]["lng"]
    return lat, lng
    
# address 중 gu 추출하는 함수
def address_to_gu(address):
    return address.split()[1]
    
# brand명이 입력되면 brand id를 반환하는 함수:
dict_cursor = conn.cursor(dictionary=True)
dict_cursor.execute("SELECT * FROM GAS_BRAND")
gas_brand = dict_cursor.fetchall()

def getBrandID(brandName):
     if brandName == '알뜰(ex)':
         brandName = '알뜰주유소'
     for item in gas_brand:
         if item['name'] == brandName:
             return item['id']
     return -1

# id 입력용 num 생성
num = 0
for index in range(len(gu_names)):
    gu_selector = f'#SIGUNGU_NM0 > option:nth-child({index+2})'
    driver.find_element(By.CSS_SELECTOR, gu_selector).click()
    station_items = driver.find_elements(By.CSS_SELECTOR,'#body1 > tr')
    
    for idx in range(len(station_items)):
        num += 1
        try:
            
            detail_selector = f'#body1 > tr:nth-child({idx+1}) > td.rlist > a'
            
            driver.find_element(By.CSS_SELECTOR, detail_selector).click()

            name = driver.find_element(By.CSS_SELECTOR, '.header #os_nm').\
            											get_attribute('innerText')
            gas = str_to_int(driver.find_element(By.CSS_SELECTOR, '#b027_p').\
            											get_attribute('innerText'))
            diesel = str_to_int(driver.find_element(By.CSS_SELECTOR, '#d047_p').\
            											get_attribute('innerText'))
            address = driver.find_element(By.CSS_SELECTOR, '#rd_addr').\
            											get_attribute('innerText')
            brand = driver.find_element(By.CSS_SELECTOR, '#poll_div_nm').\
            											get_attribute('innerText')
            
            cwsh_yn = False if '_off' in getSRC('.service #cwsh_yn') else True
            lpg_yn = False if '_off' in getSRC('.service #lpg_yn') else True
            maint_yn = False if '_off' in getSRC('.service #maint_yn') else True
            cvs_yn = False if '_off' in getSRC('.service #cvs_yn') else True
            sel24_yn = False if '_off' in getSRC('.service #sel24_yn') else True

            try:
                driver.find_element(By.CSS_SELECTOR, '#self_icon').\
                										get_attribute('alt')
                is_self = True
            except:
                is_self= False
            
            brand = brand_to_id(brand)
            gu = address_to_gu(address)
            lat,lng = address_to_latlng(address) 

            cursor.execute(sql,\
            				(num, brand,name,gu,address,gas,diesel,\
            				is_self,cwsh_yn,lpg_yn,maint_yn, cvs_yn, sel24_yn,\
                            lat, lng))
            mydb.commit()
            
            time.sleep(0.2)
        except Exception as e:
            print(e)
    time.sleep(0.5)
driver.quit()

문제 5.

시각화 프로젝트를 위하여 다음의 규칙으로 쿼리하여 CSV 파일로 저장합니다. (Python 코드로 작성)
- 전체 데이터를 가져오는데, 주유소 브랜드 아이디 대신 브랜드명이 표시되어야 합니다. (정렬 : 주유소 매장 아이디 순)
- 다음의 형식으로 저장되어야 함 (브랜드 이름, 칼럼 명 주의, id : GAS_STORE.id)

제출 5.
- 시각화 프로젝트 관련 코드 (ipynb), 결과 파일 (csv)

In [126]:


sql_query_q5 = """
SELECT GAS_STATION.id AS id, 
       GAS_BRAND.name AS brand_name, 
       GAS_STATION.name, 
       city, gu, address, 
       gasoline, diesel, self, 
       car_wash, charging_station,
       car_maintenance, convenience_store, 
       24_hours, lat, lng
FROM GAS_STATION
JOIN GAS_BRAND ON GAS_STATION.brand = GAS_BRAND.id
ORDER BY GAS_STATION.id;
"""

cur = conn.cursor()
cur.execute(sql_query_q5)
sql_result_q5 = cur.fetchall()

column_names = [desc[0] for desc in cur.description]

df = pd.DataFrame(sql_result_q5, columns=column_names)
cur.close()

df.to_csv('output.csv', index=False)



In [127]:
df = pd.read_csv('output.csv')
df.head()

id brand_name                 name city   gu                  address  \
0   1      SK에너지          (주)보성 세곡주유소   서울  강남구     서울 강남구 헌릉로 731 (세곡동)   
1   2      SK에너지              자곡셀프주유소   서울  강남구    서울 강남구 밤고개로 120 (자곡동)   
2   3   HD현대오일뱅크  HD현대오일뱅크㈜직영 산성셀프주유소   서울  강남구           서울 강남구 헌릉로 730   
3   4      GS칼텍스                방죽주유소   서울  강남구    서울 강남구 밤고개로 215 (율현동)   
4   5      SK에너지             오일프러스 셀프   서울  강남구  서울 강남구 남부순환로 2651 (도곡동)   

   gasoline  diesel  self  car_wash  charging_station  car_maintenance  \
0      1562    1439     1         1                 0                0   
1      1609    1474     1         1                 0                0   
2      1609    1464     1         1                 0                0   
3      1617    1499     1         1                 0                0   
4      1619    1524     1         1                 0                1   

   convenience_store  24_hours        lat         lng  
0                  0         0  37.467157  127.118708  
1                  0         0  37.483400  127.104046  
2                  0         0  37.466567  127.119080  
3                  0         0  37.474993  127.106887  
4                  0         1  37.485839  127.039672

문제 6.

위도, 경도 정보를 이용하여 미왕빌딩에서 1킬로 이내에 위치한 주유소 정보를 검색하세요.

- 주유소 아이디, 주유소 브랜드명, 주유소 매장명, 주소, 미왕빌딩으로 부터의 거리 (km)

제출 6.
- 쿼리, 결과

In [112]:
# 미왕빌딩 위도, 경도
# userLatitude = 37.4955366
# userHardness = 127.0293521

sql_query_q6 = """
select s.id, b.name, s.name, s.address,
         (6371*acos(cos(radians(37.4955366))*cos(radians(s.lat))*cos(radians(s.lng)
             -radians(127.0293521))+sin(radians(37.4955366))*sin(radians(s.lat))))
             distance 
       FROM GAS_BRAND b, GAS_STATION s
       where s.brand = b.id
       having distance <= 1
       order by distance;
"""

cur = conn.cursor()
cur.execute(sql_query_q6)
sql_result_q6 = cur.fetchall()
cur.close()

for row in sql_result_q6:
    print(row)


#conn.close()


(10, 'SK에너지', 'SK서광주유소', '서울 강남구 역삼로 142', 0.5234686017012263)
(259, 'HD현대오일뱅크', 'HD현대오일뱅크㈜직영 서초제일주유소', '서울 서초구 사임당로 116 (서초동)', 0.8129398651383243)
(6, 'S-OIL', '극동유화㈜ 개나리주유소', '서울 강남구 언주로 423 (역삼동)', 0.95510527469562)


문제 7.

위도, 경도 정보를 이용하여 미왕빌딩에서 셀프주유가 가능하고 24시간이면서 편의점이 있는 가장 가까운 주유소 10개를 휘 발유 가격이 가장 저렴한 순으로 정렬하여 조회하세요.

- 주유소 아이디, 주유소 브랜드명, 주유소 매장명, 주소, 휘발유 가격, 부가정보 (셀프, 24시간, 편의점 여부), 미왕빌딩으로 부터의 거리 (km)


제출 7.
- 쿼리, 결과

In [130]:
sql_query_q7 = """
SELECT s.id, b.name AS brand_name, s.name AS station_name, s.address, s.gasoline AS gasoline_price,
       CASE WHEN s.self = 1 THEN 'Yes' ELSE 'No' END AS self_service,
       CASE WHEN s.24_hours = 1 THEN 'Yes' ELSE 'No' END AS 24_hours,
       CASE WHEN s.convenience_store = 1 THEN 'Yes' ELSE 'No' END AS convenience_store,
       (6371*acos(cos(radians(37.4955366))*cos(radians(s.lat))*cos(radians(s.lng)
           -radians(127.0293521))+sin(radians(37.4955366))*sin(radians(s.lat)))) AS distance 
FROM GAS_BRAND b
JOIN GAS_STATION s ON s.brand = b.id
WHERE s.self = 1 AND s.24_hours = 1 AND s.convenience_store = 1
HAVING distance <= 10
ORDER BY s.gasoline ASC, distance ASC
LIMIT 10;
"""

cur = conn.cursor()
cur.execute(sql_query_q7)
sql_result_q7 = cur.fetchall()
cur.close()

for row in sql_result_q7:
    print(row)

(236, '알뜰(ex)', '만남의광장주유소', '서울 서초구 양재대로12길 73-71', 1528, 'Yes', 'Yes', 'Yes', 4.101206026348577)
(305, 'GS칼텍스', '지에스칼텍스(주)가든파이브주유소', '서울 송파구 탄천동로 740', 1563, 'Yes', 'Yes', 'Yes', 5.287641562631312)
(325, 'SK에너지', '백제고분로주유소', '서울 송파구 오금로 143 (방이동)', 1624, 'Yes', 'Yes', 'Yes', 7.94080482637522)
(329, 'HD현대오일뱅크', 'HD현대오일뱅크㈜직영 올림픽셀프주유소', '서울 송파구 위례성대로 188 (오금동)', 1646, 'Yes', 'Yes', 'Yes', 9.159136085042007)
(261, 'SK에너지', '대신석유(주)대우주유소', '서울 서초구 양재대로 173 (양재동)', 1679, 'Yes', 'Yes', 'Yes', 3.7080877661937928)


문제 8.

구별로 주유소 브랜드 별 휘발유 평균가격을 조회하여 저렴한 순으로 출력하세요.

- 구 이름, 주유소 브랜드 이름, 휘발유 평균 가격

제출 8.
- 쿼리, 결과

In [129]:
sql_query_q8 = """
SELECT s.gu, b.name, AVG(s.gasoline) AS avg_gasoline_price
FROM GAS_STATION s
JOIN GAS_BRAND b ON s.brand = b.id
GROUP BY s.gu, b.name
ORDER BY avg_gasoline_price ASC; 
"""

cur = conn.cursor()
cur.execute(sql_query_q8)
sql_result_q8 = cur.fetchall()
cur.close()

for row in sql_result_q8:
    print(row)

('서초구', '알뜰(ex)', Decimal('1528.0000'))
('은평구', 'HD현대오일뱅크', Decimal('1550.0000'))
('영등포구', '알뜰주유소', Decimal('1559.0000'))
('은평구', 'SK에너지', Decimal('1566.6667'))
('강북구', 'HD현대오일뱅크', Decimal('1568.3333'))
('양천구', 'S-OIL', Decimal('1569.0000'))
('강북구', 'S-OIL', Decimal('1572.0000'))
('강북구', 'GS칼텍스', Decimal('1572.3333'))
('양천구', 'GS칼텍스', Decimal('1575.0000'))
('성북구', '알뜰주유소', Decimal('1578.0000'))
('양천구', '알뜰주유소', Decimal('1578.0000'))
('서대문구', 'GS칼텍스', Decimal('1578.5000'))
('은평구', 'GS칼텍스', Decimal('1578.6667'))
('중랑구', 'HD현대오일뱅크', Decimal('1579.0000'))
('송파구', 'S-OIL', Decimal('1579.2000'))
('송파구', 'GS칼텍스', Decimal('1581.7500'))
('성북구', 'GS칼텍스', Decimal('1581.8333'))
('서대문구', 'HD현대오일뱅크', Decimal('1582.2500'))
('도봉구', 'GS칼텍스', Decimal('1582.3333'))
('중랑구', 'GS칼텍스', Decimal('1583.6667'))
('강서구', '알뜰주유소', Decimal('1584.5000'))
('성북구', 'HD현대오일뱅크', Decimal('1584.7143'))
('광진구', 'HD현대오일뱅크', Decimal('1586.1667'))
('은평구', 'S-OIL', Decimal('1587.0000'))
('성북구', 'SK에너지', Decimal('1587.6667'))
('도

---